### PR TITLE
Students Module

### PR DESCRIPTION
**Pull request by thirteenmd**
- Added pagination.

- Edit and Add Student work with Modals.

- Query Param 'page' now has the restrictions: not null, Number and bigger than 0. In case the param is one of those things there is a redirection to /students?page=1.

- Students are shown from last added to first added.

- Added Error Handling.

- On submitting the edit form, the student data is updated. (Using EventEmitter)

### CODE REVIEW

File: src/app/app.module.ts [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
-import {appRoutes} from './app.routes';
-import {HomeModule} from '@modules/home/home.module';
+import { BrowserModule } from '@angular/platform-browser';
+import { NgModule } from '@angular/core';
+import { CoreModule } from '@core/core.module';
+
+import { AppComponent } from './app.component';
+import { appRoutes } from './app.routes';
+
+import { HomeModule } from '@modules/home/home.module';
```
Comment (vadim11041996) [inline src/app/app.module.ts]: useless empty lines

File: src/app/core/core.module.ts [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
+import { CommonModule } from '@angular/common';
+import { RouterModule } from '@angular/router';
+import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
+import { InitService } from '@core/services/init/init.service';
+import { LoggerInterceptorService } from './interceptors/logger-interceptor.service';
+import { NavComponent } from './components/nav/nav.component';
+import { FooterComponent } from './components/footer/footer.component';
+import { HeaderComponent } from './components/header/header.component';
+import { LayoutComponent } from './components/layout/layout.component';
+import { CollapseModule } from 'ngx-bootstrap/collapse';
```
Comment (vadim11041996) [inline src/app/core/core.module.ts]: use empty line between angular imports and your's

File: src/app/core/services/api/backend-api.service.ts [hunk 3]
```diff
@@ -20,6 +20,7 @@ export class BackendApiService {
   };
 
   constructor(private http: HttpClient) {
+    console.log(this.URL);
```
Comment (vadim11041996) [inline src/app/core/services/api/backend-api.service.ts]: delete console.log

File: src/app/modules/students/containers/student-details-page/student-details-page.component.ts [hunk 4]
```diff
@@ -1,19 +1,53 @@
-import { Component, OnInit } from '@angular/core';
+import { Component, OnInit, OnDestroy } from '@angular/core';
 import { ActivatedRoute } from '@angular/router';
+import { StudentsService } from '@modules/students/students.service';
+import { Subscription } from '../../../../../../node_modules/rxjs';
+
+import { BsModalService } from 'ngx-bootstrap/modal';
+import { BsModalRef } from 'ngx-bootstrap/modal/bs-modal-ref.service';
+
+import Student from '../../../../shared/models/student';
```
Comment (vadim11041996) [inline src/app/modules/students/containers/student-details-page/student-details-page.component.ts]: use one empty line , and import like "@shared...."

File: src/app/modules/students/containers/student-details-page/student-details-page.component.ts [hunk 5]
```diff
@@ -1,19 +1,53 @@
...
     });
   }
 
+  ngOnDestroy() {
+    this.subscription.unsubscribe();
+  }
+
+  deleteStudent(id: number) {
+    this.studentsService.deleteStudent(id)
+      .subscribe((result) => console.log(result));
```
Comment (vadim11041996) [inline src/app/modules/students/containers/student-details-page/student-details-page.component.ts]: console

File: src/app/modules/students/containers/student-details-page/student-details-page.component.ts [hunk 6]
```diff
@@ -1,19 +1,53 @@
...
+  }
+
+  // openModal() {
+  //   this.modalRef = this.modalService.show(AddStudentModalComponent);
+  // }
```
Comment (vadim11041996) [inline src/app/modules/students/containers/student-details-page/student-details-page.component.ts]: uncoment or delete

File: src/app/modules/students/modals/add-student-modal/add-student-modal.component.html [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+               aria-describedby="idGroupHelp"
+               placeholder="Select idGroup">
+        <div *ngIf="submitted && fields.idGroup.errors"
+             class="invalid-feedback">
+          <div *ngIf="fields.idGroup.errors.required">Phone Number is required</div>
+        </div>
+      </div>
+
+      <!-- ['^[\+]?[0]{0,2}?[(]?[0-9]{3}[)]?[-\s\.]?[0-9]{3}[-\s\.]?[0-9]{4,6}$', 'i'] -->
+      <div class="form-group">
```
Comment (vadim11041996) [inline src/app/modules/students/modals/add-student-modal/add-student-modal.component.html]: delete comment

File: src/app/modules/students/modals/add-student-modal/add-student-modal.component.html [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+               type="password"
+               class="form-control"
+               [ngClass]="{ 'is-invalid': submitted && fields.password.errors }"
+               id="password"
+               autocomplete='password'
+               aria-describedby="nameHelp"
+               placeholder="Enter a password">
+        <div *ngIf="submitted && fields.password.errors"
+             class="invalid-feedback">
+          <div *ngIf="fields.password.errors.required">Password is required</div>
```
Comment (vadim11041996) [inline src/app/modules/students/modals/add-student-modal/add-student-modal.component.html]: use new variant of validation

File: src/app/modules/students/modals/add-student-modal/add-student-modal.component.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+  onSubmit() {
+    this.submitted = true;
+
+    if (this.studentForm.invalid) {
+      return;
+    }
+
+    this.api.addStudent(this.studentForm.value)
+      .subscribe((response) => console.log(response));
```
Comment (vadim11041996) [inline src/app/modules/students/modals/add-student-modal/add-student-modal.component.ts]: console

File: src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.html [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+               class="form-control"
+               [ngClass]="{ 'is-invalid': submitted && fields.phoneNumber.errors }"
+               id="phoneNumber"
+               autocomplete='phoneNumber'
+               aria-describedby="nameHelp"
+               placeholder="123-456-7890"
+               pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}">
+        <div *ngIf="submitted && fields.phoneNumber.errors"
+             class="invalid-feedback">
+          <div *ngIf="fields.phoneNumber.errors.required">Phone number is required</div>
```
Comment (vadim11041996) [inline src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.html]: same with validation function

File: package.json [hunk 11]
```diff
@@ -20,11 +20,14 @@
     "@angular/platform-browser": "^6.0.3",
     "@angular/platform-browser-dynamic": "^6.0.3",
     "@angular/router": "^6.0.3",
+    "@ng-bootstrap/ng-bootstrap": "^2.2.1",
```
Comment (alexandrubereghici) [inline package.json]: We have already ngx-boostrap,  why another library ? :)

File: src/app/modules/students/students.service.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
+import { Injectable } from '@angular/core';
+import { BackendApiService } from '@core/services/api/backend-api.service';
+import { map } from '../../../../node_modules/rxjs/operators';
+import { Observable } from '../../../../node_modules/rxjs';
+import Student from '../../shared/models/student';
```
Comment (vadim11041996) [inline src/app/modules/students/students.service.ts]: change import type

File: package.json [hunk 13]
```diff
@@ -20,11 +20,14 @@
...
     "angular-font-awesome": "^3.1.2",
     "bootstrap": "^4.1.3",
     "core-js": "^2.5.4",
     "font-awesome": "^4.7.0",
+    "ng-bootstrap": "^1.6.3",
     "ngx-bootstrap": "^3.0.1",
+    "ngx-pagination": "^3.1.1",
```
Comment (alexandrubereghici) [inline package.json]: in ngx-boostrap already exists component for pagination : https://valor-software.com/ngx-bootstrap/#/pagination

File: src/app/modules/students/containers/student-details-page/student-details-page.component.ts [hunk 14]
```diff
@@ -1,19 +1,47 @@
-import { Component, OnInit } from '@angular/core';
+import { Component, OnInit, OnDestroy } from '@angular/core';
 import { ActivatedRoute } from '@angular/router';
+import { StudentsService } from '@modules/students/students.service';
+import { Subscription } from '../../../../../../node_modules/rxjs';
+
+import { BsModalRef } from 'ngx-bootstrap/modal/bs-modal-ref.service';
+
+import Student from '../../../../shared/models/student';
```
Comment (alexandrubereghici) [inline src/app/modules/students/containers/student-details-page/student-details-page.component.ts]: fix this import . we have configured aliases.

File: src/app/modules/students/containers/student-details-page/student-details-page.component.ts [hunk 15]
```diff
@@ -1,19 +1,47 @@
-import { Component, OnInit } from '@angular/core';
+import { Component, OnInit, OnDestroy } from '@angular/core';
 import { ActivatedRoute } from '@angular/router';
+import { StudentsService } from '@modules/students/students.service';
+import { Subscription } from '../../../../../../node_modules/rxjs';
```
Comment (alexandrubereghici) [inline src/app/modules/students/containers/student-details-page/student-details-page.component.ts]: fix this import

File: src/app/modules/students/containers/student-details-page/student-details-page.component.ts [hunk 16]
```diff
@@ -1,19 +1,47 @@
...
+    private studentsService: StudentsService) { }
 
   ngOnInit() {
-    this.route.params.subscribe( params => {
-      console.log(params);
+    this.subscription = this.route.params.subscribe(params => {
+      this.id = +params['id'];
+      this.studentsService.getStudent(this.id).subscribe((element) => {
+        this.student = element;
+        this.studentsService.getStudentsGroup(this.student.idGroup).subscribe((groupName) => {
```
Comment (alexandrubereghici) [inline src/app/modules/students/containers/student-details-page/student-details-page.component.ts]: use flatmap operator for sequential observables.

File: src/app/modules/students/containers/student-details-page/student-details-page.component.ts [hunk 17]
```diff
@@ -1,19 +1,47 @@
...
+      });
     });
   }
 
+  ngOnDestroy() {
+    this.subscription.unsubscribe();
+  }
+
+  deleteStudent(id: number) {
+    this.studentsService.deleteStudent(id)
```
Comment (alexandrubereghici) [inline src/app/modules/students/containers/student-details-page/student-details-page.component.ts]: handle error messages.

File: src/app/core/core.module.ts (hunk 2 above)
Comment (thirteenmd) [reply src/app/core/core.module.ts]: *yours

File: src/app/modules/students/containers/students-page/students-page.component.ts [hunk 18]
```diff
@@ -1,15 +1,68 @@
...
+
+  constructor(private api: StudentsService,
+    private route: ActivatedRoute,
+    private modalService: BsModalService,
+    private router: Router) { }
 
   ngOnInit() {
+    this.subscription = this.route.queryParams.subscribe((params) => {
+      this.pageParam = +params.page;
+      if (this.pageParam != null || this.pageParam != NaN) {
```
Comment (alexandrubereghici) [inline src/app/modules/students/containers/students-page/students-page.component.ts]: reverse this if in order to simplify it

File: src/app/modules/students/containers/students-page/students-page.component.ts [hunk 19]
```diff
@@ -1,15 +1,68 @@
...
+
+  ngOnDestroy() {
+    this.subscription.unsubscribe();
+  }
+
+  pageChanged(selectedPage) {
+    this.page = selectedPage;
+    this.offset = this.limit * (selectedPage - 1);
+    this.subscription = this.api.getStudents(this.offset, this.limit)
+      .subscribe(students => this.students = students);
```
Comment (alexandrubereghici) [inline src/app/modules/students/containers/students-page/students-page.component.ts]: Please catch error and display it into interface.

File: src/app/modules/students/modals/add-student-modal/add-student-modal.component.ts (hunk 9 above)
Comment (alexandrubereghici) [inline src/app/modules/students/modals/add-student-modal/add-student-modal.component.ts]: Catch errors here.

File: src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.ts [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  onSubmit() {
+    this.submitted = true;
+
+    if (this.studentForm.invalid) {
+      return;
+    }
+
+    this.studentsService.updateStudentData(this.student.id, this.studentForm.value)
```
Comment (alexandrubereghici) [inline src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.ts]: catch errors

File: src/app/modules/students/students.service.ts [hunk 21]
```diff
@@ -0,0 +1,47 @@
+import { Injectable } from '@angular/core';
+import { BackendApiService } from '@core/services/api/backend-api.service';
+import { map } from '../../../../node_modules/rxjs/operators';
+import { Observable } from '../../../../node_modules/rxjs';
```
Comment (alexandrubereghici) [inline src/app/modules/students/students.service.ts]: please fix imports here.

File: src/app/modules/students/containers/student-details-page/student-details-page.component.html [hunk 22]
```diff
@@ -1,3 +1,27 @@
...
+    <h3>Last Name: {{student.lastName}}</h3>
+    <h3>Email: {{student.email}}</h3>
+    <h3>Phone Number: {{student.phoneNumber}}</h3>
+    <h3>Group Name: {{groupName}}</h3>
+
+    <app-edit-student-modal [student]="student"
+                            [parent]="this"></app-edit-student-modal>
+
+    <app-delete-student-modal [studentId]="student.id"
+                              [parent]="this"></app-delete-student-modal>
```
Comment (vadim11041996) [inline src/app/modules/students/containers/student-details-page/student-details-page.component.html]: change modal with button ,that have reference to this modal

File: src/app/app.module.ts (hunk 1 above)
Comment (crischiucucu) [reply src/app/app.module.ts]: It's not useless, it's recommended into angular style guide to separate angular and third-party imports from local imports using a line :)

File: src/app/modules/students/modals/delete-student-modal/delete-student-modal.component.ts [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+      .catch(error => {
+        this.alerts.push({ type: "danger", msg: error.message });
+        return Observable.throw(error.message);
+      })
+      .subscribe();
+    this.modalRef.hide();
+  }
+
+  openDeleteStudentModal(template: TemplateRef<any>) {
+    this.modalRef = this.modalService.show(template);
```
Comment (vadim11041996) [inline src/app/modules/students/modals/delete-student-modal/delete-student-modal.component.ts]: dont's use TemplateRef

File: src/app/app.module.ts (hunk 1 above)
Comment (thirteenmd) [reply src/app/app.module.ts]: yay ^^

File: src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.html [hunk 24]
```diff
@@ -0,0 +1,115 @@
+<button type="button"
+        class="btn btn-secondary"
+        (click)="openEditStudentModal(template)">
+  Edit
+</button>
```
Comment (vadim11041996) [inline src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.html]: remove this button

File: src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.html [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+             class="invalid-feedback">
+          <div *ngIf="formErrors('phoneNumber').required">Phone number is required</div>
+        </div>
+      </div>
+
+      <button type="submit"
+              class="btn btn-primary">Submit</button>
+    </form>
+  </div>
+</ng-template>
```
Comment (vadim11041996) [inline src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.html]: remove ng-template

File: src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.ts [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+
+  alerts: Array<any> = [];
+
+  constructor(private modalService: BsModalService,
+    private formBuilder: FormBuilder,
+    private studentsService: StudentsService) { }
+
+  ngOnInit() {
+    this.studentForm = this.formBuilder.group({
+      firstName: ['', Validators.required],
```
Comment (vadim11041996) [inline src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.ts]: don't need to build a new form, you already have one in this page

File: src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.ts [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+      lastName: ['', Validators.required],
+      email: ['', [Validators.required, Validators.email]],
+      phoneNumber: ['', Validators.required],
+      idGroup: ['', Validators.required]
+    });
+  }
+
+  openEditStudentModal(template: TemplateRef<any>) {
+    this.modalRef = this.modalService.show(template);
```
Comment (vadim11041996) [inline src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.ts]: remove this function

File: src/app/modules/students/students.module.ts [hunk 28]
```diff
@@ -1,13 +1,37 @@
...
+    StudentsPageComponent,
+    AddStudentModalComponent,
+    EditStudentModalComponent,
+    DeleteStudentModalComponent
+  ],
+  entryComponents: [
+    AddStudentModalComponent,
+    EditStudentModalComponent,
+    DeleteStudentModalComponent
+  ]
```
Comment (vadim11041996) [inline src/app/modules/students/students.module.ts]: for common components , create an array , and import it , and you are not going to write in many places the same component

File: src/app/modules/students/containers/student-details-page/student-details-page.component.ts [hunk 29]
```diff
@@ -1,19 +1,51 @@
...
   templateUrl: './student-details-page.component.html',
   styleUrls: ['./student-details-page.component.scss']
 })
-export class StudentDetailsPageComponent implements OnInit {
+export class StudentDetailsPageComponent implements OnInit, OnDestroy {
+  id: number;
+  student: Student;
+  groupName: string;
+  private subscription: Subscription;
+
```
Comment (vadim11041996) [inline src/app/modules/students/containers/student-details-page/student-details-page.component.ts]: use takeUntil instead of subsciption

File: src/app/modules/students/modals/delete-student-modal/delete-student-modal.component.html [hunk 30]
```diff
@@ -0,0 +1,23 @@
+<button type="button"
+        class="btn btn-danger"
+        (click)="openDeleteStudentModal(template)">
+  Delete
+</button>
```
Comment (vadim11041996) [inline src/app/modules/students/modals/delete-student-modal/delete-student-modal.component.html]: delete this button , use it in the parent component

File: src/app/modules/students/containers/students-page/students-page.component.html [hunk 31]
```diff
@@ -1,3 +1,46 @@
-<p>
-  students-page works!
-</p>
+<div class="container">
+  <div *ngFor="let alert of alerts">
```
Comment (alxolr) [inline src/app/modules/students/containers/students-page/students-page.component.html]: This code is in double please create a separate allert componet and reuse it

File: src/app/modules/students/modals/add-student-modal/add-student-modal.component.html [hunk 32]
```diff
@@ -0,0 +1,122 @@
+<div>
+  <div class="container">
```
Comment (alxolr) [inline src/app/modules/students/modals/add-student-modal/add-student-modal.component.html]: Create an alert component and reuse it

File: src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.html [hunk 33]
```diff
@@ -0,0 +1,115 @@
+<button type="button"
+        class="btn btn-secondary"
+        (click)="openEditStudentModal(template)">
+  Edit
+</button>
+<div class="container">
+  <div *ngFor="let alert of alerts">
+    <alert [type]="alert.type"
```
Comment (alxolr) [inline src/app/modules/students/modals/edit-student-modal/edit-student-modal.component.html]: Create reusable component

File: src/app/modules/students/containers/students-page/students-page.component.ts [hunk 34]
```diff
@@ -1,15 +1,106 @@
...
-  constructor() { }
+  constructor(private studentsService: StudentsService,
+    private route: ActivatedRoute,
+    private modalService: BsModalService,
+    private router: Router) { }
 
   ngOnInit() {
+    this.subscription = this.route.queryParams.subscribe((params) => {
+      this.pageParam = +params.page;
+      if (this.pageParam != NaN || this.pageParam != null) {
```
Comment (crischiucucu) [inline src/app/modules/students/containers/students-page/students-page.component.ts]: You can reduce the if structure like this: 
// Boolean(NaN) === false, Boolean(null) === false and Boolean(0) === false; so if pageparam will have one of this //value it will coerce to false so do like this:
if(this.pageParam) {
this.setPage(this.pageParam);
} else {
this.setPage(1);
}

File: src/app/modules/students/containers/students-page/students-page.component.ts [hunk 35]
```diff
@@ -1,15 +1,106 @@
...
+        }
+      } else {
+        this.setPage(1)
+      }
+    });
+
+    this.subscription = this.studentsService.getNumberOfStudents()
+      .pipe(
+        mergeMap((studentsNumber) => {
+          this.totalItems = +studentsNumber;
```
Comment (crischiucucu) [inline src/app/modules/students/containers/students-page/students-page.component.ts]: You could use this.offset = studentsNumber - 10; (- operator will do implicit coercion to number)

File: src/app/modules/students/modals/add-student-modal/add-student-modal.component.html [hunk 36]
```diff
@@ -0,0 +1,122 @@
...
+  </div>
+  <div class="modal-body">
+    <form [formGroup]="studentForm"
+          (submit)="onSubmit()">
+      <div class="form-group">
+        <label for="firstName">First name</label>
+        <input formControlName="firstName"
+               type="text"
+               class="form-control"
+               [ngClass]="{ 'is-invalid': isSubmitted && formErrors('firstName') }"
```
Comment (crischiucucu) [inline src/app/modules/students/modals/add-student-modal/add-student-modal.component.html]: Please ctreate specified getters instead of formErrors() method, ex: get firstName() {} and use it into template firstName.errors

File: src/app/modules/students/modals/add-student-modal/add-student-modal.component.html [hunk 37]
```diff
@@ -0,0 +1,122 @@
...
+          <div *ngIf="formErrors('firstName').required">First Name is required</div>
+        </div>
+      </div>
+
+      <div class="form-group">
+        <label for="lastName">Last name</label>
+        <input formControlName="lastName"
+               type="text"
+               class="form-control"
+               [ngClass]="{ 'is-invalid': isSubmitted && formErrors('lastName') }"
```
Comment (crischiucucu) [inline src/app/modules/students/modals/add-student-modal/add-student-modal.component.html]: The syntax for binding class using directive ngClass it's ok, but if you have a single class to bind please use the [class.is-invalid]="isSubmitted && formErrors('lastName')"

File: src/app/modules/students/modals/add-student-modal/add-student-modal.component.html [hunk 38]
```diff
@@ -0,0 +1,122 @@
...
+               id="lastName"
+               autocomplete='lastName'
+               aria-describedby="nameHelp"
```
Comment (crischiucucu) [inline src/app/modules/students/modals/add-student-modal/add-student-modal.component.html]: If you don't have the element with the id="nameHelp" please don't use aria-describedby attr

File: src/app/modules/students/containers/students-page/students-page.component.ts [hunk 39]
```diff
@@ -1,15 +1,106 @@
...
+    }
+
+    this.subscription = this.studentsService.getStudents(this.offset, this.limit)
+      .catch(error => {
+        this.alerts.push({ type: "danger", msg: error.message });
+        return Observable.throw(error.message);
+      })
+      .subscribe((students) => {
+        this.students = students;
+        this.students.reverse();
```
[the remark this answers is not in the document]
Comment (thirteenmd) [reply src/app/modules/students/containers/students-page/students-page.component.ts]: To display the students from last added to first added. Vadim suggested doing that.

File: src/app/modules/students/modals/add-student-modal/add-student-modal.component.html [hunk 40]
```diff
@@ -0,0 +1,122 @@
...
+        <label for="phoneNumber">Phone Number</label>
+        <input formControlName="phoneNumber"
+               type="tel"
+               class="form-control"
+               [ngClass]="{ 'is-invalid': isSubmitted && formErrors('phoneNumber') }"
+               id="phoneNumber"
+               autocomplete='phoneNumber'
+               aria-describedby="nameHelp"
+               placeholder="123-456-7890"
+               pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}">
```
[the remark this answers is not in the document]
Comment (thirteenmd) [reply src/app/modules/students/modals/add-student-modal/add-student-modal.component.html]: All the phone numbers in the database have been generated according to this format.

File: src/app/modules/students/containers/students-page/students-page.component.ts (hunk 39 above)
[the remark this answers is not in the document]
Comment (vadim11041996) [reply src/app/modules/students/containers/students-page/students-page.component.ts]: If we added a new student, to see it, we need to go to the last page, but so, we'll see it instantly

File: src/app/modules/students/containers/students-page/students-page.component.ts [hunk 41]
```diff
@@ -1,15 +1,106 @@
...
+          return this.studentsService.getStudents(this.offset, this.limit);
+        })
+      )
+      .catch(error => {
+        this.alerts.push({ type: "danger", msg: error.message });
+        return Observable.throw(error.message);
+      })
+      .subscribe((students) => {
+        this.students = students;
+        this.students.reverse();
```
[the remark this answers is not in the document]
Comment (vadim11041996) [reply src/app/modules/students/containers/students-page/students-page.component.ts]: we want to show the item in the reverse form, so the last that's added, to be shown as the first  item

File: src/app/shared/services/paginator-helper/paginator-helper.service.ts [hunk 42]
```diff
@@ -0,0 +1,29 @@
+import { Injectable } from '@angular/core';
+import { Observable } from '../../../../../node_modules/rxjs';
+
```
Comment (vadim11041996) [inline src/app/shared/services/paginator-helper/paginator-helper.service.ts]: change import type

File: src/app/modules/students/containers/students-page/students-page.component.ts [hunk 43]
```diff
@@ -1,15 +1,114 @@
...
   templateUrl: './students-page.component.html',
   styleUrls: ['./students-page.component.scss']
 })
-export class StudentsPageComponent implements OnInit {
+export class StudentsPageComponent implements OnInit, OnDestroy {
+  defaultItemsNumber: number = 10;
 
-  constructor() { }
+  offset: number = 0;
+  limit: number = this.defaultItemsNumber;
```
Comment (vadim11041996) [inline src/app/modules/students/containers/students-page/students-page.component.ts]: why do you use `limit` and  `defaultItemsNumber` ?

File: src/environments/environment.prod.ts [hunk 44]
```diff
@@ -1,4 +1,4 @@
 export const environment = {
   production: true,
-  apiUrl: '10.177.106.10:80'
+  apiUrl: '10.177.106.10:80/api/v1'
```
Comment (vadim11041996) [inline src/environments/environment.prod.ts]: this is not going to work ,  it should be   http://10.177.106.10:80/api/v1

Comment (thirteenmd) [reply src/environments/environment.prod.ts]: Yeah, it is ```http://10.177.106.10:80/api/v1```.

File: src/app/modules/students/containers/students-page/students-page.component.ts (hunk 43 above)
Comment (thirteenmd) [reply src/app/modules/students/containers/students-page/students-page.component.ts]: Cristina suggested I use a variable for the default number of items in a page, which is 10, because I use it in more than one place.

File: src/app/modules/students/students.module.ts [hunk 45]
```diff
@@ -1,13 +1,39 @@
...
 import { StudentDetailsPageComponent, StudentsPageComponent } from '@modules/students/containers';
+import { DeleteStudentModalComponent } from './modals/delete-student-modal/delete-student-modal.component';
+import { SharedModule } from '@shared/shared.module';
+
 
+const COMPONENTS = [
+  AddStudentModalComponent,
+  EditStudentModalComponent,
+  DeleteStudentModalComponent,
+];
```
Comment (crischiucucu) [inline src/app/modules/students/students.module.ts]: Add a space above the class decorator

File: src/app/modules/students/students.service.ts [hunk 46]
```diff
@@ -0,0 +1,49 @@
...
+import { Observable } from 'rxjs';
+
+import { BackendApiService } from '@core/services/api/backend-api.service';
+import { Student } from '@shared/models/student';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class StudentsService {
+  id: number;
```
Comment (crischiucucu) [inline src/app/modules/students/students.service.ts]: Why do you need the id as a service property, usually it is send as arguments in  methods.

File: src/app/modules/students/students.service.ts [hunk 47]
```diff
@@ -0,0 +1,49 @@
...
+  constructor(private api: BackendApiService) { }
+
+  getNumberOfStudents() {
```
Comment (crischiucucu) [inline src/app/modules/students/students.service.ts]: Add return type please. Verify all the service methods to have a return type :)

File: src/app/modules/students/students.service.ts [hunk 48]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  getStudent(id: number): Observable<Student> {
+    return this.api.get(`students/${id}`)
+      .pipe(map(result => result));
+  }
+
+  addStudent(newStudent): Observable<any> {
+    return this.api.post(`students`, newStudent)
+      .pipe(map(result => result));
```
Comment (crischiucucu) [inline src/app/modules/students/students.service.ts]: here you also don't need a pipe() just return this.api.post(`students`, newStudent);

File: src/app/modules/students/students.service.ts [hunk 49]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  updateStudentData(id: number, editedStudent) {
+    return this.api.put(`students/${id}`, editedStudent)
+      .pipe(map(result => console.log(result)));
```
Comment (crischiucucu) [inline src/app/modules/students/students.service.ts]: you should return the response from put for catching errors outside

File: src/app/modules/students/students.service.ts [hunk 50]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  deleteStudent(id: number) {
+    return this.api.delete(`students/${id}`)
+      .pipe(map(result => result));
```
Comment (crischiucucu) [inline src/app/modules/students/students.service.ts]: you don't need a pipe() just  return this.api.delete(`students/${id}`);

File: src/app/shared/componets/alert/local-alert/local-alert.component.ts [hunk 51]
```diff
@@ -7,7 +7,7 @@ import { Alert, AlertType } from '../../../models/alert';
   templateUrl: './local-alert.component.html',
 })
 export class LocalAlertComponent {
-  @Input() alerts: Alert[] = [];
+  @Input('alerts') alerts: Alert[] = [];
```
Comment (crischiucucu) [inline src/app/shared/componets/alert/local-alert/local-alert.component.ts]: Why did you add an alias @Input('alerts')? The name of the property it's the same as the alias. The alias is used usualy to change the name of the property outside. Ex: @Input('alerts') allAlerts: [];

File: src/app/shared/models/student.ts [hunk 52]
```diff
@@ -1,10 +1,10 @@
 import { User } from '@shared/models/user';
 
 export class Student extends User {
-  groupId: number;
+  idGroup: number;
```
Comment (crischiucucu) [inline src/app/shared/models/student.ts]: Please verify well here, it seems that someone renamed the properties and you rename them like in the previous version.

File: src/environments/environment.prod.ts (hunk 44 above)
Comment (crischiucucu) [inline src/environments/environment.prod.ts]: Please don't commit this file. The environements should be changed just in case it really changed.

File: src/app/modules/students/containers/students-page/students-page.component.ts [hunk 53]
```diff
@@ -1,15 +1,114 @@
...
+        this.setPage(1)
+      }
+    });
+  }
+
+  initNumberOfStudents() {
+    this.studentsService.getNumberOfStudents()
+      .pipe(
+        mergeMap((studentsNumber) => {
+          this.totalItems = +studentsNumber;
```
[the remark this answers is not in the document]
Comment (thirteenmd) [reply src/app/modules/students/containers/students-page/students-page.component.ts]: totalItems is not a pagination param. It's just the total number of students.

File: src/app/modules/students/containers/students-page/students-page.component.ts [hunk 54]
```diff
@@ -1,15 +1,114 @@
...
+import { Alert, AlertType } from '@shared/models/alert';
+import { STUDENTS_URL } from '@shared/constants';
 @Component({
   selector: 'app-students-page',
   templateUrl: './students-page.component.html',
   styleUrls: ['./students-page.component.scss']
 })
-export class StudentsPageComponent implements OnInit {
+export class StudentsPageComponent implements OnInit, OnDestroy {
+  defaultItemsNumber = 10;
```
Comment (crischiucucu) [inline src/app/modules/students/containers/students-page/students-page.component.ts]: Please look if there is a constant in constants with this value, I think I saw it elsewhere, if not, please add it as a constant and use it.

File: src/app/modules/students/containers/students-page/students-page.component.ts [hunk 55]
```diff
@@ -1,15 +1,114 @@
...
+        takeUntil(this.destroy$),
+        catchError(
+          (error) => {
+            this.alerts.push({ type: AlertType.Error, message: error });
+            return throwError(error);
+          })
+      );
+  }
+
+  reverseStudents(students) {
```
Comment (crischiucucu) [inline src/app/modules/students/containers/students-page/students-page.component.ts]: I think it's better to name this function like setStudents() and do first the revert, after asign to this.students, because you modify the property twice and it will init change detection cycle. Ex:
setStudents() {
  this.students = students.reverse();
}

File: src/app/modules/students/modals/add-student-modal/add-student-modal.component.ts [hunk 56]
```diff
@@ -0,0 +1,93 @@
...
+})
+export class AddStudentModalComponent implements OnInit, OnDestroy {
+  studentForm: FormGroup;
+  isSubmitted = false;
+
+  alerts: Alert[] = [];
+
+  destroy$: Subject<boolean> = new Subject<boolean>();
+
+  public event: EventEmitter<any> = new EventEmitter();
```
Comment (crischiucucu) [inline src/app/modules/students/modals/add-student-modal/add-student-modal.component.ts]: Please rename the event into more suggestive identifier, ex: studentAdded or other name that is appropriate. And you don't have to specify public in front of the property declaration because it's by default public when the specifier is missing(in typescript)
